R/install.R

```diff
@@ -1,276 +1,343 @@
+      ## TODO: update progress bar here
+      task <- select_next_task(state)
+      state <- start_task(state, task)
+    }
 
-  running <- character()
-  results <- list()
-  processes <- list()
-  update_progress <- function(count) {
-    bar$tick(count, tokens = list(packages = collapse(running, ", "), processes = length(processes), num_workers = num_workers))
-  }
-  update_progress(0)
-
-  # Currently assumes the order of installation is not important
-  events <- get_events(list(plan = plan, processes = list(), results = list()), num_workers, lib, bar)
-  while(running_processes(events)) {
-    for (output in get_output(events)) {
-      is_running <- grepl("^Building", strip_style(output$output))
-      running <- union(running, sub("^Building ", "", strip_style(output$output[is_running])))
-      finished <- output$output[!is_running]
-
-      if (length(finished) > 0 && nzchar(finished)) {
-
-        # remove installed from running
-        installed <- strip_style(finished)[grepl("^. Installed ", strip_style(finished))]
-        installed <- sub(". Installed ([^[:space:]]+).*", "\\1", installed)
-
-        running <- setdiff(running, installed)
-        bar$message(finished)
-        update_progress(length(installed))
-      }
-      if (!bar$finished && length(running)) {
-        update_progress(0)
-      }
+    while (1) {
+      if (are_we_done(state)) break;
+      ## TODO: update progress bar here
+
+      events <- poll_workers(state)
+      state <- handle_events(state, events)
+      task  <- select_next_task(state)
+      state <- start_task(state, task)
     }
-    events <- get_events(events, num_workers, lib, bar)
+  }, error = function(e) kill_all_processes(state))
+
+  create_install_result(state)
+}
+
+make_start_state <- function(plan, config) {
+
+  ## We store the data about build and installation here
+  install_cols <- data.frame(
+    stringsAsFactors = FALSE,
+    build_done = plan$type == "installed" | plan$binary,
+    build_time = ifelse(plan$binary, 0, NA_real_),
+    build_error = I(rep_list(nrow(plan), list())),
+    build_stdout = I(rep_list(nrow(plan), character())),
+    build_stderr = I(rep_list(nrow(plan), character())),
+    install_done = plan$type == "installed",
+    install_time = ifelse(plan$type == "installed", 0, NA_real_),
+    install_error = I(rep_list(nrow(plan), list())),
+    install_stdout = I(rep_list(nrow(plan), character())),
+    install_stderr = I(rep_list(nrow(plan), character())),
+    worker_id = NA_character_
+  )
+  plan <- cbind(plan, install_cols)
+
+  list(
+    plan = plan,
+    workers = list(),
+    config = config)
+}
+
+are_we_done <- function(state) {
+  all(state$plan$install_done)
+}
+
+#' @importFrom callr poll
+
+poll_workers <- function(state) {
+  if (length(state$workers)) {
+    timeout <- get_timeout(state)
+    procs <- lapply(state$workers, "[[", "process")
+    res <- poll(procs, ms = timeout)
+    vapply(res, function(x) "ready" %in% x, logical(1))
```

Comment:
We can either use [map_lgl()](https://github.com/r-lib/pkginstall/blob/064dcbcd4b6d0fa1855c7fb541819ecb215fc2f6/R/utils.R#L88) or remove `map_lgl()` and define `vlapply()` in the utils. Or just depend on purrr and use `purrr::map_lgl()`

Reply:
We could, if we are already depending on rlang.

Reply:
Btw. what is the reason for this?
```
map_lgl <- get("map_lgl", asNamespace("rlang"))
```
This makes rlang an install-time dependency.